src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -1386,6 +1413,37 @@ public IList<string> ValidValues
         {
             get
             {
+                if (validValuesGeneratorType != null)
+                {
+                    ValidValuesCacheEntry validValuesCacheEntry;
+                    if (validValuesCache.TryGetValue(validValuesGeneratorType, out validValuesCacheEntry))
+                    {
+                        var currentTime = DateTime.Now;
+                        if (DateTime.Compare(validValuesCacheEntry.cacheEntryCreationTime.AddSeconds(CacheExpiration), currentTime) < 0)
+                        {
+                            // Update expired Entry.
+                            validValuesCacheEntry.validValues = validValuesCacheEntry.validValuesGenerator.GetValidValues()?.ToArray();
+                            validValuesCacheEntry.cacheEntryCreationTime = currentTime;
+                            validValuesCache[validValuesGeneratorType] = validValuesCacheEntry;
+                        }
+                    }
+                    else
+                    {
+                        var validValuesGenerator = (IValidateSetValuesGenerator)Activator.CreateInstance(validValuesGeneratorType);
+                        var validValues = validValuesGenerator.GetValidValues()?.ToArray();
+                        var cacheEntryCreationTime = DateTime.Now;
+                        validValuesCacheEntry = new ValidValuesCacheEntry(validValuesGenerator, validValues, cacheEntryCreationTime);
+                        validValuesCache.TryAdd(validValuesGeneratorType, validValuesCacheEntry);
+                    }
+                    if (validValuesCacheEntry.validValues == null)
+                    {
+                        throw new ValidationMetadataException(
+                            "ValidateSetGeneratedValidValuesListIsEmpty",
+                            null,
+                            Metadata.ValidateSetGeneratedValidValuesListIsEmpty);
+                    }
+                    _validValues = validValuesCacheEntry.validValues;
```

Comment:
I think it would also be useful to clean up the cache - this could be done with something like `Task.Delay(CacheExpiration).ContinueWith(() => validValuesCacheEntry.validValues = null);`

This might not seem like a big deal, the collection should typically be small, but sometimes it might be large, so we should make the collection collectible if it's truly going to be used again.